blockchain/chain_sync/src/sync.rs

```diff
@@ -153,33 +186,157 @@ where
+                        if i < batch_size {
+                            batch_size = i;
+                        }
+
+                        // set params for blocksync request
+                        let idx = i - batch_size;
+                        let next = &ts[usize::try_from(idx)?];
+                        let req_len = u64::try_from(batch_size + 1)?;
+
+                        // only fetch messages
+                        const MESSAGES: u64 = 2;
+                        // receive tipset bundle from block sync
+                        let ts_bundle = self
+                            .network
+                            .blocksync_request(
+                                peer_id.clone(),
+                                BlockSyncRequest {
+                                    start: next.cids().to_vec(),
+                                    request_len: req_len,
+                                    options: MESSAGES,
+                                },
+                            )
+                            .await?;
+
+                        for b in ts_bundle.chain {
+                            // construct full tipsets from fetched messages
+                            let fts = self.tipset_msgs(
+                                ts[usize::try_from(i)?].clone(),
+                                &b.bls_msgs,
+                                &b.secp_msgs,
+                                b.bls_msg_includes.len(),
+                                b.secp_msg_includes.len(),
+                            )?;
+                            // validate tipset and messages
+                            self.validate_tipsets(fts)?;
+                            // store messages
+                            self.chain_store.put_messages(&b.bls_msgs)?;
+                            self.chain_store.put_messages(&b.secp_msgs)?;
+                        }
+                    }
+                    i -= REQUEST_WINDOW;
+                    continue;
+                }
+            };
+            // full tipset found in storage; validate and continue
+            self.validate_tipsets(fts)?;
+            i -= 1;
+            continue;
+        }
+
+        Ok(())
+    }
+
+    /// Returns FullTipset with unchecked messages
+    fn tipset_msgs(
+        &self,
+        ts: Tipset,
+        bls_msgs: &[UnsignedMessage],
+        secp_msgs: &[SignedMessage],
+        bls_msg_len: usize,
+        secp_msg_len: usize,
+    ) -> Result<FullTipset, Error> {
+        // see https://github.com/filecoin-project/lotus/blob/master/build/params_shared.go#L109 for block message limit
+        const BLOCK_MESSAGE_LIMIT: usize = 512;
+        let mut blocks: Vec<Block> = Vec::with_capacity(ts.blocks().len());
+
+        if ts.blocks().len() != secp_msg_len || ts.blocks().len() != bls_msg_len {
+            return Err(Error::Other(
+                "Msg included length didnt match tipset size".to_string(),
+            ));
+        }
```

Comment:
This check makes no sense, please explain

Reply:
See lotus: https://github.com/filecoin-project/lotus/tree/master/https://github.com/filecoin-project/lotus/blob/master/chain/sync.go#L287

Reply:
yeah? those last two parameters are the 2d arrays for message includes indexes, you are assigning ALL messages to ALL blocks, when you should be using the indexes to attach the propoer messages. That check you are doing here makes no sense